[vcpkg/fortran] Switch to ifort as the Fortran compiler

Neumann-A opened the pull request:
- [x] Depends on https://github.com/microsoft/vcpkg-tool/pull/148 being merged
- [x] waiting for next vcpkg-tool release
~- [ ]  #19554 for uwp builds~
- Use helper port.
- ~~Update Lapack to 3.10~~ extracted 
- **No longer limited to dynamic linkage for Fortran**
- No x86 host support.

(will update versions if this can actually correctly run in CI)

also closes  #19983

Neumann-A committed:
add fortran helper port

Neumann-A committed:
update lapack port and switch to new fortran helper

Neumann-A committed:
[skip ci] cleanup

Neumann-A committed:
[skip ci] add vcpkg-ifort to copy dlls + more cleanup

Neumann-A commented on ports/vcpkg-fortran/z_vcpkg_load_environment_from_batch.cmake:
```diff
@@ -0,0 +1,86 @@
+### BATCH_FILE_PATH
+Arguments to pass to the batch file
+#]===]
+function(z_vcpkg_load_environment_from_batch)
```
@strega-nil @ras0219-msft : Should I put this into `scripts/cmake` ? The only reason I did not do it here is because I didn't want to force a world rebuild until the required changes to vcpkg-tool are merged. 
I also need this for the community intel triplet. (also to load the required env variables)

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Neumann-A committed:
[skip actions] apply format diff

Neumann-A committed:
[skip actions] consider 19 20 and 21 and a bit of cleanup

Neumann-A committed:
[skip actions] fix vcpkg-ifort

Neumann-A committed:
[skip actions] make it work in static builds

Neumann-A commented:
@BillyONeal: VMs not completely setup?

```
    The license for Visual Studio expires in 1 day.
    Project 'ALL_BUILD' could not be loaded because it's missing install components. To fix this launch Visual Studio installer with the following selections:
    C++ (v142) Universal Windows Platform tools
```

BillyONeal commented:
> @BillyONeal: VMs not completely setup?

Nothing has changed in the VM image in the last month (https://github.com/microsoft/vcpkg/pull/19508 isn't landed yet). Note that we install "Universal Windows Platform Tools" but not "Universal Windows Platform Tools (v142)" right now

https://github.com/microsoft/vcpkg/blob/eb9cb291669f24b315d6001065e3b1e53e3701b4/scripts/azure-pipelines/windows/deploy-visual-studio.ps1#L11

This implies installing the "Universal Windows Platform Development" workload, but I don't see lines for the following optional checkboxes (they are checked on my development workstation but not CI :) ):

![image](https://user-images.githubusercontent.com/1544943/129395815-8016c2db-70d2-400a-9bcc-eed18e988863.png)

Neumann-A commented:
According to https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2019 the missing optional component is `Microsoft.VisualStudio.ComponentGroup.UWP.VC.BuildTools`

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Neumann-A committed:
[skip actions] fix uwp builds.

Neumann-A committed:
[skip actions] apply format diff

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Neumann-A committed:
Merge branch 'use_oneapi_fortran' of https://github.com/Neumann-A/vcpkg into use_oneapi_fortran

Neumann-A committed:
[skip actions] update vcpkg-fortran to the changes made in rollup

Neumann-A committed:
[skip actions] switch to version-date and revert changes to the portfile

Neumann-A committed:
[skip actions] version stuff

Neumann-A committed:
add lapack-reference:x64-uwp=skip on the ci baseline until VM is updated.
(If UWP would ninja this would probably not be a VM update issue)

Neumann-A committed:
revisit portfile of vcpkg-ifort

Neumann-A committed:
overwrite version

Neumann-A committed:
fix dependency stuff.
Fortran dlls are not build with App Container Bit !
(probably needs libs from llvm flang in the future)

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Neumann-A committed:
fix manifest stuff

Neumann-A committed:
version stuff

Neumann-A committed:
fix supports field

Neumann-A committed:
update version

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/lapack-reference/portfile.cmake
#	ports/lapack-reference/vcpkg.json
#	versions/baseline.json
#	versions/l-/lapack-reference.json

Neumann-A committed:
update version stuff lost in merge

ras0219-msft commented on ports/vcpkg-fortran/x_vcpkg_find_fortran.cmake:
```diff
@@ -0,0 +1,69 @@
+                message(FATAL_ERROR "Batch file to setup Intel oneAPI not found! Please provide a correct ONEAPI_ROOT and make sure it contains setvars.bat!")
+            endif()
+            z_vcpkg_load_environment_from_batch(BATCH_FILE_PATH "${SETVARS}")
+            if(VCPKG_TARGET_IS_UWP)
+                set(extra_uwp_flags "/NODEFAULTLIB /Qopenmp-stubs /D_UNICODE /DUNICODE /DWINAPI_FAMILY=WINAPI_FAMILY_APP /D__WRL_NO_DEFAULT_LIB__")
+                set(exta_uwp_link_flags "-DCMAKE_SHARED_LINKER_FLAGS_INIT:STRING=/APPCONTAINER")
+            endif()
```
We could edit the DLLs after the fact and add the bit -- though it may fail WACK.

BillyONeal replied:
It isn't just adding the bit; there's also the whole onecore vs. kernel32 thing... :(

PhoebeHui commented on scripts/ci.baseline.txt:
```diff
@@ -1716,6 +1716,7 @@ clapack:x64-windows-static-md=skip
 clapack:x64-windows=skip
 clapack:x86-windows=skip
 clapack:x64-uwp=skip
+
```
```suggestion
```

PhoebeHui commented:
Limits to relate knowledge, requires other team member further review.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/lapack/vcpkg.json
#	ports/vcpkg-cmake/vcpkg.json
#	versions/baseline.json
#	versions/l-/lapack.json
#	versions/v-/vcpkg-cmake.json

Neumann-A committed:
version stuff

Neumann-A committed:
remove ws change in baseline

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

BillyONeal commented:
The requested component is included in VMs as of 2921-12-14

Neumann-A commented:
@BillyONeal I thought the result was that UWP is not supported by Fortran since the required fortran runtime libraries don't link against UWP stuff?

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/vcpkg-cmake/vcpkg.json
#	versions/baseline.json
#	versions/v-/vcpkg-cmake.json

Neumann-A committed:
version stuff

Neumann-A committed:
Merge branch 'use_oneapi_fortran' of https://github.com/Neumann-A/vcpkg into use_oneapi_fortran

BillyONeal commented:
> @BillyONeal I thought the result was that UWP is not supported by Fortran since the required fortran runtime libraries don't link against UWP stuff?

Oh maybe that's true/likely. I didn't follow the history etc; I just cleaned out all the `depends:vm-update` requested changes in the December Patch Tuesday.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/vcpkg-cmake/vcpkg.json
#	versions/baseline.json
#	versions/v-/vcpkg-cmake.json

Neumann-A committed:
version stuff

Commit:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/vcpkg-cmake/vcpkg.json
#	versions/baseline.json
#	versions/v-/vcpkg-cmake.json

Commit:
add license stuff

Commit:
version stuff

Commit:
make helper ports use host dependencies

Commit:
version stuff

Commit:
remove wrong dep

Commit:
version stuff

Commit:
remove deprecated function

Commit:
version stuff

Commit:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/lapack-reference/portfile.cmake
#	ports/lapack/vcpkg.json
#	versions/l-/lapack-reference.json
#	versions/l-/lapack.json

Neumann-A committed:
version stuff

Commit:
fix license file

Commit:
v stuff

Commit:
debug stuff

BillyONeal commented:
> @BillyONeal: The VM update seems to have removed `IFORT_COMPILER21` and `ONEAPI_ROOT` from the environment.

We updated the oneapi installer indicated before to one that supports VS2022 and perhaps the current one no longer sets those?

Neumann-A commented:
@BillyONeal : As far as I can tell those still get set by the new installers. I removed the environment variables and reinstalled the newer version and the variables reappeared.

BillyONeal commented:
> @BillyONeal : As far as I can tell those still get set by the new installers. I removed the environment variables and reinstalled the newer version and the variables reappeared.

When I rerun the deploy scripts now the link appears broken :sigh:. I guess we need to start storing these internally like we do the CUDAnn bits :(

JackBoosY commented:
@BillyONeal Any plan now?

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	ports/vcpkg-cmake/vcpkg.json
#	versions/baseline.json
#	versions/v-/vcpkg-cmake.json

Neumann-A commented:
@BillyONeal you can remove the tag "depends:vm- update". Seems like ifort is again working. Only some old baseline regressions are left over.

BillyONeal commented:
> @BillyONeal you can remove the tag "depends:vm- update". Seems like ifort is again working.

Yay!

JackBoosY commented:
The baseline regression was fixed by https://github.com/microsoft/vcpkg/pull/24169.

Commit:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

JackBoosY commented:
Will rerun x64-osx tests after others finished.

Neumann-A committed:
version stuff

Commit:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

Commit:
detect intel fortran compiler to apply patch in lapack.

Commit:
version stuff

Commit:
Merge remote-tracking branch 'upstream/master' into use_oneapi_fortran

# Conflicts:
#	versions/v-/vcpkg-cmake.json

Commit:
version date bump

Commit:
version stuff

BillyONeal commented:
Related: https://github.com/microsoft/vcpkg/pull/15053 ?

Neumann-A commented:
> Related: https://github.com/microsoft/vcpkg/pull/15053 ?

So so. 15053 switches the whole windows toolchain to intel. 15053 contains stuff from this PR though.

BillyONeal commented on ports/lapack-reference/FindLAPACK.cmake:
```diff
@@ -8,11 +8,11 @@ FindLAPACK
 Find Linear Algebra PACKage (LAPACK) library
```
I did not review this file but it appears to be copy-pasta out of a CMake distribution; if there are specific changes you think needed to be pointed out can you indicate them? (It's really unfortunate that we don't just directly use CMake's copy here...)

Neumann-A replied:
Dropped from the other PR.

BillyONeal commented on ports/lapack-reference/portfile.cmake:
```diff
@@ -8,17 +8,24 @@ if(EXISTS "${CURRENT_INSTALLED_DIR}/share/clapack/copyright")
 
 set(lapack_ver 3.10.0)
 
+set(VCPKG_CRT_LINKAGE_BACKUP ${VCPKG_CRT_LINKAGE})
```
Should this use vcpkg_backup_env_variables and vcpkg_restore_env_variables ?

Neumann-A replied:
yes probably. I think those functions did not exist when the PR was created.

Neumann-A replied:
wait this is not an environment variable. This is a triplet variable.

BillyONeal commented on ports/vcpkg-fortran/z_vcpkg_load_environment_from_batch.cmake:
```diff
@@ -0,0 +1,86 @@
+    )
+    file(READ "${CURRENT_BUILDTREES_DIR}/environment-after-out.log" ENVIRONMENT_AFTER)
+    
+    # Escape characters that have a special meaning in CMake strings.
+    string(REPLACE "\\" "/"     ENVIRONMENT_INITIAL "${ENVIRONMENT_INITIAL}")
```
The comment and this line don't agree since this will affect paths and similar. The correct escape would be `"\\\\"`? (Or the comment should be fixed)

Neumann-A replied:
No the intention here is to transform `[[\]]` -> `[[/]]` and its intention is to change paths. Otherwise the whole script falls apart. I tried different stuff to keep raw `\` (e.g. `[[\]]` -> `[[\\]]` or not doing the replacement) but it doesn't seem to work. The compiler is not found which indicates some cmake magic going on with he handling of `\`. Since windows doesn't care about `\` vs `/` it just transforms everything to `/` since `\` has special meaning in cmake. Just like `;` and `\n` is transformed to `;` to make it a list for the foreach. 

This is also the reason why `file(STRINGS)` with `REGEX` cannot be used here since EOL `\` in a path like variable will be interpreted as a line continuation.

Neumann-A replied:
This is still required. Since this is intended and the comment itself is correct I mark this as resolved. I hope using [[\]] makes it more obvious that it intention is to transform all `\`

BillyONeal reviewed (changes requested):
* Our default behavior has to work without the user installing extra Intel stuff for users who do not care. This probably prevents using ifort by default since it appears to require a separate login-walled installation of Intel's OneAPI toolchains.
* Capturing a port for the contract of "we need Fortran support", as you have done here with `vcpkg-fortran`, seems reasonable and consistent with other things that we do. (I'm personally concerned about potential command line differences between them but I'm told that lapack and blas are substantially all our Fortran customers and that I probably shouldn't worry about that)
  * We do think it's reasonable to give users a mechanism to choose a Fortran compiler to use, as you are trying to demonstrate in https://github.com/microsoft/vcpkg/pull/15053 . A triplet setting, as demonstrated there, would work, but is inefficient because it forces a world rebuild to change this even though a small number of ports may be affected. (Of course, this is true for all triplet settings) Overlays are a solution that affects only things that actually need Fortran, but providing an overlay forces the user to implement the complete interface somehow rather than just saying "I want ifort" or "I want gfortran". Overlays have a nice property in that it lets you do development on the "not default" option without triggering world rebuilds for other customers.
  * We do not think "try to use ifort if we can find it" is OK, it needs to be an explicit user decision, for example, in the triplet.
  * This contract/metaport probably shouldn't have a "supports" clause, unless we know that it's categorically impossible to provide Fortran there. So `!uwp` may be fine but `windows` is not. (Linux customers may want to use gfortran from their system's package manager, for example.)
* [You mentioned this but I'm indicating that we're considering it] This PR breaks x86 hosts (you need to be an amd64 host in order to run this). Some years ago we had a substantial number of x86-only customers. We accidentially broke this ourselves though in https://github.com/microsoft/vcpkg/pull/19365/ and didn't switch back to CMake to extract things for 4 ish months and it seems nobody noticed which suggests that this wouldn't be a big deal.
* Given that UWP is already broken for gfortran et al. here, don't worry about UWP interaction in whatever the solution is.


I'm sorry if these observations sound a bit "wishy washy" -- the situation is that we aren't 100% sure what we want the future to be here so we've resorted to listing constraints we believe exist on the solution space to help you decide (a) how you should interpret our response as to whether we want this, and (b) whether you want to continue investigating this path.

Suggested path forward:

1. Add `vcpkg-fortran` as the representation of the contract, as you have done here. Its behavior should match today's behavior (gfortran).
2. Document the contract (all the cmake functions, variables, environment variables, etc.) expected to be provided by `vcpkg-fortran`.
3. Create a new port intended to be an overlay for `vcpkg-fortran` which selects ifort. That probably needs to live in a different directory like the hypothetical "backports" we have been talking about, but for purposes of making forward progress on this PR I don't think you need a concrete answer to that question yet.

Neumann-A commented on ports/vcpkg-cmake/vcpkg_cmake_build.cmake:
```diff
@@ -58,11 +58,13 @@ function(vcpkg_cmake_build)
-        set(parallel_args "/m")
+        if(NOT VCPKG_USE_INTERNAL_Fortran)
+            set(build_args
+                "/p:VCPkgLocalAppDataDisabled=true"
```
CMake in Fortran mode using the VS Generator does not use `MSBuild` but `devenv` something where these flags are unknown and error.

Neumann-A replied:
removed from https://github.com/microsoft/vcpkg/pull/24622. This was mainly required for UWP being tied to MSBuild but since that has been solved and figured out that it is impossible to correctly build for UWP currently I dropped this change

Neumann-A commented:
> Our default behavior has to work without the user installing extra Intel stuff for users who do not care. This probably prevents using ifort by default since it appears to require a separate login-walled installation of Intel's OneAPI toolchains.

With that you are basically saying to take the flang-classic route used in #23987. The main reason I am hesitant about that route is the build times required for the LLVM fork of flang-classic and the fact that #23987 loses x86 target which can still be covered by gfortran. Would be good if that could be reduced somehow

> lapack and blas are substantially all our Fortran customers

that is because the Fotran support situation has not been solved and vcpkg is focused around windows. On other platforms Fortran support is easily available. Only windows is a problem. 

> A triplet setting, as demonstrated there, would work, but is inefficient because it forces a world rebuild to change

It is not a triplet setting but a toolchain setting which sets `CMAKE_Fortran_(COMPILER|FLAGS)` explicitly. This is only required on windows due to vcpkg cleaning the environment/path. On other platforms CMake will find the fortran compiler without extra help and without needing extra flags. 
Getting rid of the world rebuild due to the triplet hash is an open issue #19984. 
The only thing which would be nice to have is `z_vcpkg_load_environment_from_batch` to not have to manually setup the passthrough. Should this be made into a different PR?

>  rather than just saying "I want ifort" or "I want gfortran". Overlays have a nice property in that it lets you do development on the "not default" option without triggering world rebuilds for other customers.

Personally I would like to use `version-string` for that in the main repo ;). basically have a PR with `version-string : "gfortran"`; `version-string : "ifort"`, `version-string : "flang-classic"`, The additional overlays for classic users could also be provided. The question here is what should the end default be in the main repo?
Again this is a windows only issues. All other platforms will control the Fortran compiler via env FC. 

> This contract/metaport probably shouldn't have a "supports" clause, unless we know that it's categorically impossible to provide Fortran there. So !uwp may be fine but windows is not. (Linux customers may want to use gfortran from their system's package manager, for example.)

All other platforms have native fortran support and don't require the contract/extra port. I don't intend to force the compiler there. 

> ...This PR breaks x86 hosts .... We accidentially broke this ourselves ... nobody noticed which suggests that this wouldn't be a big deal.....  Given that UWP is already broken for gfortran et al. here, don't worry about UWP interaction in whatever the solution is.

Ok won't care about x86 host and UWP. This also means I should go with #23987 which additionally adds arm64 support but loses x86 target support. I could make a port `vcpkg-fortran` depending on nothing (for !windows) and depending on `vcpkg-gfortran` for x86 or `vcpkg-flang-classic` for the rest. 

>the situation is that we aren't 100% sure what we want the future to be here 

I personally think that should be clear -> `flang-new` will be the future path which probably will still require 2 years to mature.

BillyONeal commented:
>With that you are basically saying to take the flang-classic route used in #23987. The main reason I am hesitant about that route is the build times required for the LLVM fork of flang-classic and the fact that #23987 loses x86 target which can still be covered by gfortran. Would be good if that could be reduced somehow

Yeah, we wouldn't want to require people build a whole compiler to get BLAS. But clang is available in binary forms today, it seems like it would be reasonable for flang to be similarly available. (It may even involve some amount of hosting by GitHub or the vcpkg team in support of this or...)

> Personally I would like to use version-string for that in the main repo ;).

Then we would need to have version versions! It'd be turtles all the way down! I would point to the "alternatives" / "options" RFC as a solution here whenever that finally gets to the top of our backlog.

>>This contract/metaport probably shouldn't have a "supports" clause, unless we know that it's categorically impossible to provide Fortran there. So !uwp may be fine but windows is not. (Linux customers may want to use gfortran from their system's package manager, for example.)
>All other platforms have native fortran support and don't require the contract/extra port. I don't intend to force the compiler there.

The port should still exist to document the dependency even if it becomes empty on the platforms where Fortran compilers are assumed to be "ambiently available".

Neumann-A commented:
> But clang is available in binary forms today, it seems like it would be reasonable for flang to be similarly available.

It is not. Also building the flang1 and flang2 doesn't require that much time. Only building clang with the classig-flang toolchain takes forever (1h on my machine). Due to MT(d) MD(d) stuff there needs to be extra stuff taken into account in the building of clang with flang on windows. I haven't figured out how clang embeds `/DEFAULTLIB: ` into object files. 
It probably gets better if flang-new is finally ready to produce binaries maybe it will then be included in the LLVM windows package but currently only ifort/ifx is available in binary form on windows (unless you plan to pull flang from conda.)

> Then we would need to have version versions!

that already exists: It is called `port-version:`

>  finally gets to the top of our backlog.

Probably when we are all grey and wiser ;)

> The port should still exist to document the dependency even if it becomes empty on the platforms where Fortran compilers are assumed to be "ambiently available".

probably can do that and just assume all other platforms use gfortran even though there is ifort and ifx also available.

Neumann-A commented on ports/vcpkg-fortran/copyright:
```diff
@@ -0,0 +1,23 @@
+Copyright (c) Microsoft Corporation
```
It is: `${VCPKG_ROOT_DIR}/LICENSE.txt`.

But yes done in the other PR

Neumann-A commented on ports/vcpkg-fortran/portfile.cmake:
```diff
@@ -0,0 +1,16 @@
+set(FUNCTION_NAME x_vcpkg_find_fortran)
+
+if(VCPKG_CROSSCOMPILING)
+    # make FATAL_ERROR in CI when issue #16773 fixed
+    message(WARNING "${PORT} is a host-only port; please mark it as a host port in your dependencies.")
```
In https://github.com/microsoft/vcpkg/pull/24622 `vcpkg-fortran` cannot be a native only port. (Reasons are found in that PR)
